rust/azure_iot_operations_connector/src/file_mount_azure_device_registry.rs

```diff
@@ -1,4 +1,1110 @@
+    ///
+    /// Returns Ok([`DeviceEndpointCreateObservation`]) if the observation is successfully created, otherwise returns an [`Error`].
+    ///
+    /// # Arguments
+    /// * `debounce_duration` - The duration to debounce incoming I/O events.
+    ///
+    /// # Errors
+    /// - [`Error`] of kind [`ErrorKind::EnvironmentVariableError`] if the environment variable was not able to be read.
+    /// - [`Error`] of kind [`ErrorKind::WatcherError`] if the watcher could not be created.
+    /// - [`Error`] of kind [`ErrorKind::IoError`] if there are issues accessing the file mount.
+    /// - [`Error`] of kind [`ErrorKind::ParseError`] if there are issues parsing the file names and content.
+    pub fn new(debounce_duration: Duration) -> Result<DeviceEndpointCreateObservation, Error> {
+        let mount_path = get_mount_path()?;
+
+        // This channel is used to send notifications about device endpoint creation
+        let (create_device_tx, create_device_rx) = mpsc::unbounded_channel();
+
+        // Tracks devices and assets in the file mount.
+        let mut file_mount_map = FileMountMap::new(create_device_tx.clone());
+
+        // Add the already existing device endpoints and their assets to the file mount map
+        get_device_endpoint_names(&mount_path)?
+            .iter()
+            .try_for_each(|device| {
+                file_mount_map.insert_device_endpoint(device);
+
+                let assets = get_asset_names(&mount_path, device)?;
+                file_mount_map.update_assets(device, &assets);
+
+                Ok::<_, Error>(())
+            })?;
+
+        // TODO: When the number of files being tracked is large, the watcher might not be able to keep up with the events.
+        // In the future we should consider adding redundancy checks (like hashing device endpoint files)
+        // to ensure that the file mount map is in sync with the file mount.
+        let mut debouncer = new_debouncer(
+            debounce_duration,
+            None,
+            move |res: Result<Vec<notify_debouncer_full::DebouncedEvent>, Vec<notify::Error>>| {
+                match res {
+                    Ok(events) => {
+                        // TODO: There might be a case where we receive events out of order (i.e a modify before a create). The file mount map accounts for this but
+                        // it might be better to just match on an Any.
+                        for debounced_event in &events {
+                            match debounced_event.event.kind {
+                                EventKind::Create(event::CreateKind::File) => { // Event signals the creation of a device endpoint
+                                    debounced_event.paths.iter().for_each(|path| {
+                                        // TODO: We could use an expect here since we are sure the path is valid
+                                        let Some(mount_path) = path.parent() else {
```

Comment:
is this just a safeguard check? Because we know what this value should be already, right?

Reply:
That's a good point, this should not be different from the one we started watching